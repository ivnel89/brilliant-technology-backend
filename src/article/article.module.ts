import { Module } from '@nestjs/common';
import { ArticleService } from './article.service';
import { ArticleController } from './article.controller';
import { TypeOrmModule } from '@nestjs/typeorm';
import { UserModule } from 'src/user/user.module';
import { Article } from './entities/article.entity';

@Module({
  imports: [
    TypeOrmModule.forFeature([Article]),
    UserModule
  ],
  controllers: [ArticleController],
  providers: [ArticleService],
  exports: [ArticleService]
})
export class ArticleModule {}
