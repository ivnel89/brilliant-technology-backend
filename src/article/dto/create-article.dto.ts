import { ApiProperty } from "@nestjs/swagger";
import { IsString, IsUUID } from "class-validator";

export class CreateArticleDto {
    @ApiProperty()
    @IsString()
    content: string;
  
    @ApiProperty()
    @IsString()
    @IsUUID()
    authorId: string;
}
