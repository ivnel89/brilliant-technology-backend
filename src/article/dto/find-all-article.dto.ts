import { ApiProperty } from "@nestjs/swagger";
import { IsNumber, IsOptional, IsUUID, Max } from "class-validator";

export class FindAllArticleDto {
    @IsOptional()
    @IsNumber()
    @ApiProperty({
      required: false
    })
    offset?: number;

    @IsOptional()
    @IsNumber()
    @Max(250)
    @ApiProperty({
      required: false
    })
    limit?: number;

    @IsOptional()
    @ApiProperty({
      required: false
    })
    @IsUUID()
    requesterId?: string;
  }
  