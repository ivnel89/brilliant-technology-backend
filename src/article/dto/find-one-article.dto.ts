import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsUUID } from "class-validator";

export class FindOneArticleDto {
    @IsOptional()
    @ApiProperty({
      required: false
    })
    @IsUUID()
    requesterId?: string;
  }
  