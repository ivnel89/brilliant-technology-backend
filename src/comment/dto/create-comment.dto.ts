import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsString, IsUUID, Length } from "class-validator";

export class CreateCommentDto {
  @ApiProperty()
  @IsString()
  @Length(1, 10000)
  content: string;

  @ApiProperty()
  @IsString()
  @IsUUID()
  authorId: string;

  @ApiProperty()
  @IsString()
  @IsUUID()
  articleId: string;

  @ApiProperty({
    nullable: true,
    required: false
  })
  @IsString()
  @IsUUID()
  @IsOptional()
  parentCommentId: string;
}
