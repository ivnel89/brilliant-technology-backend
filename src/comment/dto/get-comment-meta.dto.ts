import { ApiProperty } from "@nestjs/swagger";
import { IsArray, IsString, IsUUID } from "class-validator";

export class GetCommentsDto {
  @ApiProperty()
  @IsString()
  @IsUUID()
  userId: string;

  @ApiProperty({
     isArray: true
  })
  @IsArray()
  commentIds: Array<string>;
}
