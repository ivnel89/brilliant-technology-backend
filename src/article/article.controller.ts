import { Controller, Get, Post, Body, Patch, Param, Delete, Query } from '@nestjs/common';
import { ArticleService } from './article.service';
import { CreateArticleDto } from './dto/create-article.dto';
import { FindAllArticleDto } from './dto/find-all-article.dto';

@Controller('article')
export class ArticleController {
  constructor(private readonly articleService: ArticleService) {}

  @Post()
  create(@Body() createArticleDto: CreateArticleDto) {
    return this.articleService.create(createArticleDto);
  }

  @Get()
  findAll(
    @Query() findAllArticleDto?: FindAllArticleDto,
  ) {
    return this.articleService.findAll(findAllArticleDto);
  }

  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.articleService.findOne(id);
  }  
}
