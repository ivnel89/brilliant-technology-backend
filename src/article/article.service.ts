import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { UserService } from 'src/user/user.service';
import { Repository } from 'typeorm';
import { CreateArticleDto } from './dto/create-article.dto';
import { FindAllArticleDto } from './dto/find-all-article.dto';
import { Article } from './entities/article.entity';

@Injectable()
export class ArticleService {
  constructor(
    @InjectRepository(Article)
    private articleRepository: Repository<Article>,
    private readonly userService: UserService
  ) {}

  async create(createArticleDto: CreateArticleDto) {
    const user = await this.userService.findOne(createArticleDto?.authorId);
    return this.articleRepository.save(
      new Article(user, createArticleDto?.content),
    );
  }

  findAll(findAllArticleDto: FindAllArticleDto): Promise<Array<Article>> {
    const { limit = 10, offset = 0} = findAllArticleDto
    return this.articleRepository.find({
      order: {
        lastModifiedDate: 'ASC',
      },
      skip: offset,
      take: limit,
    });  }

  findOne(id: string): Promise<Article> {
    return this.articleRepository.findOne({
      where: {
        id,
      },
    });  
  }
}
