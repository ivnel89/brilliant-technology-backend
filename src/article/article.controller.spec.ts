import { Test, TestingModule } from '@nestjs/testing';
import { plainToInstance } from 'class-transformer';
import { validate } from 'class-validator';
import { UserRepository } from 'src/user/test/user.repository.mock';
import { UserService } from 'src/user/user.service';
import { ArticleController } from './article.controller';
import { ArticleService } from './article.service';
import { CreateArticleDto } from './dto/create-article.dto';
import { FindAllArticleDto } from './dto/find-all-article.dto';
import { createArticleDto, findAllArticleDto, findOneArticleDto, mockArticleId } from './test/article.dto.mock';
import { ArticleRepository } from './test/article.repository.mock';

describe('ArticleController', () => {
  let controller: ArticleController;
  let service: ArticleService;
  let serviceCreateSpy,
    serviceFindAllSpy,
    serviceFindOneSpy

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [ArticleController],
      providers: [
        ArticleService,
        ArticleRepository,
        UserService,
        UserRepository,
      ],
    }).compile();

    controller = module.get<ArticleController>(ArticleController);
    service = module.get<ArticleService>(ArticleService);
    serviceCreateSpy = jest.spyOn(service, 'create');
    serviceFindAllSpy = jest.spyOn(service, 'findAll'); 
    serviceFindOneSpy = jest.spyOn(service, 'findOne'); 
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  describe('create', () => {
    const serviceCreateError = "serviceCreateError";
    it('should call create from articleService with right params', async () => {
      await expect(controller.create(createArticleDto)).resolves.toMatchSnapshot().then(() => {
        expect(serviceCreateSpy).toBeCalledWith(createArticleDto);
      })
    })
    it('should validate createArticleDto', async () => { 
      const invalidInput = new CreateArticleDto();
      invalidInput.authorId = 999 as any;
      invalidInput.content = undefined;
      const errors = await validate(plainToInstance(CreateArticleDto,invalidInput))
      expect(errors.length).toBe(2)
    })
    it('should throw an error when create from articleService throws an error', async () => {
      serviceCreateSpy.mockRejectedValueOnce(serviceCreateError);
      await expect(controller.create(createArticleDto)).rejects.toBe(
        serviceCreateError,
      );
    })
  })

  describe('findAll', () => {
    const serviceFindAllError = "serviceFindAllError";
    it('should call findAll from articleService with right params', async () => {
      await expect(controller.findAll(findAllArticleDto)).resolves.toMatchSnapshot().then(() => {
        expect(serviceFindAllSpy).toBeCalledWith(findAllArticleDto);
      })
    })
    it('should validate findAllArticleDto', async () => { 
      const invalidInput = new FindAllArticleDto();
      invalidInput.limit = NaN;
      invalidInput.offset = "twentyTwo" as any;
      const errors = await validate(plainToInstance(FindAllArticleDto,invalidInput))
      expect(errors.length).toBe(2)
    })
    it('should throw an error when findAll from articleService throws an error', async () => {
      serviceFindAllSpy.mockRejectedValueOnce(serviceFindAllError);
      await expect(controller.findAll(findAllArticleDto)).rejects.toBe(
        serviceFindAllError,
      );
    })
  })

  describe('findOne', () => {
    const serviceFindOneError = "serviceFindOneError";
    it('should call findOne from articleService with right params', async () => {
      await expect(controller.findOne(mockArticleId, findOneArticleDto)).resolves.toMatchSnapshot().then(() => {
        expect(serviceFindOneSpy).toBeCalledWith(mockArticleId);
      })
    })
    it('should throw an error when findOne from articleService throws an error', async () => {
      serviceFindOneSpy.mockRejectedValueOnce(serviceFindOneError);
      await expect(controller.findOne(mockArticleId, findOneArticleDto)).rejects.toBe(
        serviceFindOneError,
      );
    })
  })
});
