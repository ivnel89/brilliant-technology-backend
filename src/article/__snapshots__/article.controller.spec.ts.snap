// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleController create should call create from articleService with right params 1`] = `
ArticleContract {
  "author": UserContract {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "id": undefined,
    "lastName": "Bar",
  },
  "content": "Foo Bar",
  "createdDate": undefined,
  "id": undefined,
}
`;

exports[`ArticleController findAll should call findAll from articleService with right params 1`] = `
Array [
  ArticleContract {
    "author": UserContract {
      "displayPicture": "www.example.com/img.png",
      "firstName": "Foo",
      "id": undefined,
      "lastName": "Bar",
    },
    "comments": Array [
      CommentContract {
        "author": UserContract {
          "displayPicture": "www.example.com/img.png",
          "firstName": "Foo",
          "id": undefined,
          "lastName": "Bar",
        },
        "content": "comment content",
        "createdDate": undefined,
        "id": undefined,
        "upVoted": false,
      },
    ],
    "content": "article content",
    "createdDate": undefined,
    "id": undefined,
  },
]
`;

exports[`ArticleController findOne should call findOne from articleService with right params 1`] = `
ArticleContract {
  "author": UserContract {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "id": undefined,
    "lastName": "Bar",
  },
  "comments": Array [
    CommentContract {
      "author": UserContract {
        "displayPicture": "www.example.com/img.png",
        "firstName": "Foo",
        "id": undefined,
        "lastName": "Bar",
      },
      "content": "comment content",
      "createdDate": undefined,
      "id": undefined,
    },
  ],
  "content": "article content",
  "createdDate": undefined,
  "id": undefined,
}
`;
